pkg/types/validation/installconfig.go

```diff
@@ -99,6 +102,122 @@ func ValidateInstallConfig(c *types.InstallConfig, openStackValidValuesFetcher o
 	if _, ok := validPublishingStrategies[c.Publish]; !ok {
 		allErrs = append(allErrs, field.NotSupported(field.NewPath("publish"), c.Publish, validPublishingStrategyValues))
 	}
+
+	return allErrs
+}
+
+// IPAddressTypeByField is a map of field path to whether they request IPv4 or IPv6.
+type IPAddressTypeByField map[string]struct{ IPv4, IPv6 bool }
+
+// IPByField is a map of field path to the net.IPs in sorted order.
+type IPByField map[string][]net.IP
+
+// InferIPVersionFromInstallConfig infers the user's desired ip version from the networking config.
+// Presence field names match the field path of the struct within the Networking type. This function
+// assumes a valid install config.
+func InferIPVersionFromInstallConfig(n *types.Networking) (hasIPv4, hasIPv6 bool, presence IPAddressTypeByField, addresses IPByField) {
```

Comment:
does is need to be public?

Reply:
AWS and Azure asset gen (specifically tfvars) would use it.  I can move it to another package but none of them were clearly downstream of types.  Prefer it to be out of validation but didn’t have a good target.  Don’t want

Reply:
> AWS and Azure asset gen (specifically tfvars) would use it.

this function has a very big API as it provides so many more details than I think tfvars generation would require.. maybe expose smaller API?

Reply:
I may not need the whole thing for TFVars, but there needs to be an unambiguous signal of ipv6 vs ipv4 vs dual stack, and that's half of this method (for instance, on AWS dual stack you don't need an ipv6 machine cidr).  I suppose given our current restrictions we could check via service network, so that might be the minimal data we need in assets.TFVars (I will link the commit that proposes what TFvars change would be for all platforms for singel/dual stack in a few)

Reply:
I'm going to move this private, detection in tfvars in https://github.com/openshift/installer/pull/2847/commits/bb3272fd932f90bb0f84edd72ae3ae6120293adf is what handles this now